pkg/server/plugin/nodeattestor/tpmdevid/challenge.go

```diff
@@ -0,0 +1,87 @@
+package tpmdevid
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpm2/credactivation"
+	devid "github.com/spiffe/spire/pkg/common/plugin/tpmdevid"
+)
+
+// We use a 32 bytes nonce to provide enough cryptographical randomness and to be
+// consistent with other nonces sizes around the project.
+const nonceSize = 32
+
+func newDevIDChallenge() ([]byte, error) {
+	nonce := make([]byte, nonceSize)
+	_, err := rand.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	return nonce, nil
+}
+
+func verifyDevIDChallenge(cert *x509.Certificate, challenge, response []byte) error {
+	return cert.CheckSignature(cert.SignatureAlgorithm, challenge, response)
+}
+
+func newCredActivationChallenge(akPub, ekPub tpm2.Public) (*devid.CredActivation, []byte, error) {
+	akName, err := akPub.Name()
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot extract name from AK public")
+	}
+
+	hash, err := ekPub.NameAlg.Hash()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	nonce := make([]byte, hash.Size())
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	encKey, err := ekPub.Key()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var symBlockSize int
+	switch encKey.(type) {
+	case *rsa.PublicKey:
```

Comment:
Do we not support non-RSA ek certs/keys?

Reply:
Currently the agent read the EK certificate from the NV index defined for RSA 2048 EK certificates (0x01c00002). There is another NV index defined for ECC NIST P256 EK certificates (0x01c0000a). Those are called the [low-rage(section 2.2.1.4)](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p3_r2_pub.pdf#page=12) NV indexes.

Do you think we should also check for the ECC EK certificate in this first version of the plugin? It could be configurable or we could detect the first EK certificate index that is populated and use that one.


PS: additionally, there is another set of indexes called the [high-range(section 2.2.1.5)](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p3_r2_pub.pdf#page=13) that define EK data for other algorithms. Probably we don't want to include high-range this yet.